### PR TITLE
TimerTask: kill the execution when it exceeds the task timeout 

### LINKED ISSUE #639: Thread leaks in TimerTask for long-executing jobs
For long-executing jobs and small `execution_interval` values, `TimerTask` will leak threads at the rate of `execution_interval`.

`timeout_interval`, whilst documented, doesn't seem to be referenced anywhere. There seems to be an old (but unmerged) PR (https://github.com/ruby-concurrency/concurrent-ruby/pull/526) that at least solves the problem of timeouts running at the rate of `execution_interval` instead of `timeout_interval`.

But that still wouldn't solve the problem of a long-running task still being rescheduled every `timeout_interval`, even if it couldn't be explicitly killed off. Nor can `TimerTask` be run without a timeout interval, if you were happy to just let tasks run until they're complete and reschedule them for another period afterwards.

### PR DESCRIPTION
**Pull request by rubemz**
Before this commit when a TimerTask task exceeded the timeout the job kept running. That could lead to thread leaking. Related to #639

This PR changes the TimerTask executor to be a SingleThreadExecutor. So whenever the task doesn't finish in time, the execution thread is killed (in this case the pool is killed).

The spec ensures that this new behavior is working correctly. It fails on master. If the main job isn't killed after the timeout, the `latch.wait(2)` would raise, since the timer task sleeps for 5 seconds.

If you want to see the leak by yourself take a look on this https://github.com/rubemz/concurrent-ruby-leak, which is pretty much a copy from https://github.com/ruby-concurrency/concurrent-ruby/issues/639#issuecomment-309792568. @shun0309 thanks for that code.

Closes #639

### CODE REVIEW

File: lib/concurrent/timer_task.rb [hunk 1]
```diff
@@ -325,6 +339,10 @@ def execute_task(completion)
     def timeout_task(completion)
       return unless @running.true?
       if completion.try?
+        @executor.kill
+        @executor.wait_for_termination
+        @executor = Concurrent::RubySingleThreadExecutor.new()
```
Comment (fabiokr) [inline lib/concurrent/timer_task.rb]: Maybe just `new`?

Comment (boone) [reply lib/concurrent/timer_task.rb]: (also line 284)

**Comment by pitr-ch** [comment]
Thanks for the fix, I'll have a look in couple of days.

File: lib/concurrent/timer_task.rb (hunk 1 above)
Comment (rubemz) [reply lib/concurrent/timer_task.rb]: 61e0074

**Comment by pitr-ch** [comment]
Hi, sorry I did not forgot about your PR. I needed to get out 1.1.0.pre1 for testing. I'll review the PR soon.

File: lib/concurrent/timer_task.rb [hunk 2]
```diff
@@ -280,7 +280,8 @@ def ns_initialize(opts, &task)
       self.execution_interval = opts[:execution] || opts[:execution_interval] || EXECUTION_INTERVAL
       self.timeout_interval = opts[:timeout] || opts[:timeout_interval] || TIMEOUT_INTERVAL
       @run_now = opts[:now] || opts[:run_now]
-      @executor = Concurrent::SafeTaskExecutor.new(task)
+      @task = task
+      @executor = Concurrent::RubySingleThreadExecutor.new
```
[the remark this answers is not in the document]
Comment (rubemz) [reply lib/concurrent/timer_task.rb]: A

[the remark this answers is not in the document]
Comment (rubemz) [reply lib/concurrent/timer_task.rb]: @wjordan thanks. fixed in 9e2fd1c

**Comment by pitr-ch** [comment]
Sorry still did not find time to do a proper review. There is a problem with killing threads, if it's killed while running e.g. ensure block it may corrupt data and other unpleasant things (more info e.g. https://www.mikeperham.com/2015/05/08/timeout-rubys-most-dangerous-api/). I need to give this a more careful thought.

**Pull request closed without merging**